### PR TITLE
Refine HttpChannel.Listener

### PR DESCRIPTION
**Pull request by gregw**
addendum to #9754 
better comments
split write into two cheaper events

### CODE REVIEW

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannelListeners.java [hunk 1]
```diff
@@ -45,6 +45,7 @@ public class HttpChannelListeners implements HttpChannel.Listener
 
     private static final MethodType LISTENER_TYPE_ON_RESPONSE_COMMITTED = methodType(Void.TYPE, Request.class, Integer.TYPE, HttpFields.class);
     private static final MethodType LISTENER_TYPE_ON_RESPONSE_WRITE = methodType(Void.TYPE, Request.class, Boolean.TYPE, ByteBuffer.class);
+    public static final MethodHandle EMPTY_ON_RESPONSE_WRITE_HANDLE = MethodHandles.empty(LISTENER_TYPE_ON_RESPONSE_WRITE);
```
Comment (joakime) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannelListeners.java]: Don't do this, you risk pinning a classloader using `MethodHandles.empty()` as a static final.

Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannelListeners.java]: @joakime Huh? This class already has plenty of static final `MethodHandle`s that are on actual classes?  How can an empty handle pin a classloader?

Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannelListeners.java]: @joakime I changed to use a boolean anyway.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java [hunk 2]
```diff
@@ -128,95 +128,116 @@ public interface HttpChannel extends Invocable
      * <p>Listener instances that are set as a bean on the {@link Connector} are
      * efficiently added to {@link HttpChannel}.
      */
-    public interface Listener extends EventListener
+    interface Listener extends EventListener
     {
         /**
-         * Invoked just after the HTTP request line and headers have been parsed.
+         * Invoked just after the HTTP request line and headers have been parsed
+         * (i.e. from within the call to {@link HttpChannel#onRequest(MetaData.Request)}).
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:135]: Wrap with `<p>`.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java [hunk 3]
```diff
@@ -128,95 +128,116 @@ public interface HttpChannel extends Invocable
...
          * This the state of the request from the network, and does not include
          * any request customizations (eg: forwarding, secure, etc)
          * </p>
          *
-         * <p>
-         * Implementations must not block!
-         * </p>
-         *
-         * @param request the request object
+         * @param request the request object. The {@code read()} and {@code demand(Runnable)} methods must not be called by the listener.
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:142]: I think calling `read()` and `demand()` should not be a worry -- if called wrongly it's the user fault.

**Review by sbordet — CHANGES_REQUESTED** [review]

**Review by sbordet — COMMENTED** [review]
Partial review, but we have discussed this online.

**Pull request merged**